### PR TITLE
[Fix #3994] Add new cop  checks for invalid arguments for find_by

### LINKED ISSUE #3994: New cop that check if find_by(*) works fine
I'd like to suggest a new cop, which detects misusing of `find_by` method (ie: `find_by(some_id)`) and points out it should be `find_by(id: some_id)`
I think it's really hard to find that misusing.

Note: I'm not sure if this is rubocop's business, so if it's not then tell me so please. Thanks :)

### PR DESCRIPTION
**Pull request by konumashuri**
close #3994 
This cop will check if `find_by` arguments are either `id: some_id` or `{ id: some_id }`.
This will check each args with `args.all? { |arg| arg.is_a?(Rubocop::AST::HashNode) }`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/cop/rails/find_by_arg.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      #   User.find_by(1)
+      #
+      #   # good
+      #   User.find_by(id: 1)
+      class FindByArg < Cop
+        MSG = '`find_by(arg)` may not work. Use `find_by(column: arg)` instead.'.freeze
+
+        def on_send(node)
+          _receiver, method_name, *args = *node
+          return unless method_name == :find_by
```
Comment (pocke) [inline lib/rubocop/cop/rails/find_by_arg.rb]: Can you support `find_by!` (with bang) ?

File: lib/rubocop/cop/rails/find_by_arg.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+          return if args.all? { |arg| arg.is_a?(RuboCop::AST::HashNode) }
```
Comment (pocke) [inline lib/rubocop/cop/rails/find_by_arg.rb]: We can use `arg.hash_type?` for this purpose.

File: lib/rubocop/cop/rails/find_by_arg.rb (hunk 1 above)
Comment (konumashuri) [inline lib/rubocop/cop/rails/find_by_arg.rb]: Sure!

File: lib/rubocop/cop/rails/find_by_arg.rb (hunk 2 above)
Comment (konumashuri) [inline lib/rubocop/cop/rails/find_by_arg.rb]: I didn't know that, thank you.

**Comment by bbatsov** [comment]
I think we can improve this cop further if we raised an alarm simply if `find_by` is passed something with a name `id`, `something_id` or some non-hash literal. In its current state the cop is going to produce too many false positives.

File: lib/rubocop/cop/rails/find_by_arg.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      # change them to use `find_by(column: arg)` instead.
+      #
+      # @example
+      #   # bad
+      #   User.find_by(1)
+      #
+      #   # good
+      #   User.find_by(id: 1)
+      class FindByArg < Cop
+        MSG = '`find_by(arg)` may not work.' \
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/find_by_arg.rb]: I'd change this message to something like `Use `find_by(column: arg)` or `find` instead.` or something similar.

**Comment by bbatsov** [comment]
@konumashuri ping :-)

**Comment by bbatsov** [comment]
@konumashuri ping again :-)

**Comment by bbatsov** [comment]
Closing due to lack of activity.

**Pull request closed without merging**